README
NAME
    WebService::SQLFormat - Format SQL via the sqlformat.org API

VERSION
    version 0.000005

SYNOPSIS
        use strict;
        use warnings;
        use feature qw( say );

        use WebService::SQLFormat;
        my $formatter = WebService::SQLFormat->new(
            identifier_case => 'upper',
            reindent        => 1,
        );

        my $sql = shift @ARGV;

        say $formatter->format_sql($sql);

DESCRIPTION
    BETA BETA BETA. Subject to change.

    This module is a thin wrapper around <http://sqlformat.org>

AUTHOR
    Olaf Alders <olaf@wundercounter.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2016 by Olaf Alders.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for WebService-SQLFormat

{{$NEXT}}

0.000007  2017-03-06 11:11:24-05:00 America/Toronto
    - Adds explicit dependency on LWP::Protocol::https

0.000006  2016-10-04 16:06:18-04:00 America/Toronto
    - sqlformat.org has switched to SSL. Reported by Slaven Rezić
      https://github.com/oalders/webservice-sqlformat/issues/2
    - Now including actual docs.

0.000005  2016-07-05 11:07:49-04:00 America/Toronto
    - Disable debugging in tests unless it's an author test.

0.000004  2016-06-30 16:44:33-04:00 America/Toronto

0.000003  2016-06-30 16:09:46-04:00 America/Toronto
    - Renames sqlformat script to sql-webformat.

0.000002  2016-06-30 15:45:10-04:00 America/Toronto
    - Add sqlformat script.
    - Return the result key from format_sql() rather than a hashref.

0.000001  2016-06-21 17:05:43-04:00 America/Toronto
    - Initial release.


lib/WebService/SQLFormat.pm
package WebService::SQLFormat;

use Moo 2.002004;

use JSON::MaybeXS qw( decode_json );
use LWP::UserAgent ();
use Module::Runtime qw( use_module );
use Types::Standard qw( Bool InstanceOf Int Str );
use Types::URI qw( Uri );

has debug_level => (
    is      => 'ro',
    isa     => Int,
    default => 0,
);

has identifier_case => (
    is            => 'ro',
    isa           => Str,
    predicate     => '_has_identifier_case',
    documentation => q{'upper', 'lower' or 'capitalize'},
);

has keyword_case => (
    is            => 'ro',
    isa           => Str,
    predicate     => '_has_keyword_case',
    documentation => q{'upper', 'lower' or 'capitalize'},
);

has reindent => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);

has strip_comments => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);

has ua => (
    is      => 'ro',
    isa     => InstanceOf ['LWP::UserAgent'],
    lazy    => 1,
    builder => '_build_ua',
);

has url => (
    is      => 'ro',
    isa     => Uri,
    coerce  => 1,
    lazy    => 1,
    default => 'https://sqlformat.org/api/v1/format',
);

sub _build_ua {
    my $self = shift;
    my $ua   = LWP::UserAgent->new;
    return $ua unless $self->debug_level;

    use_module( 'LWP::ConsoleLogger::Easy', 0.000028 );
    LWP::ConsoleLogger::Easy::debug_ua( $ua, $self->debug_level );
    return $ua;
}

sub format_sql {
    my $self = shift;
    my $sql  = shift;

    my $res = $self->ua->post(
        $self->url,
        {
            (
                $self->_has_identifier_case
                ? ( identifier_case => $self->identifier_case )
                : ()
            ),
            (
                $self->_has_keyword_case
                ? ( keyword_case => $self->keyword_case )
                : ()
            ),
            reindent       => $self->reindent,
            sql            => $sql,
            strip_comments => $self->strip_comments,
        }
    );
    return decode_json( $res->decoded_content )->{result};
}

1;
__END__

# ABSTRACT: Format SQL via the sqlformat.org API

=head1 DESCRIPTION

BETA BETA BETA.  Subject to change.

This module is a thin wrapper around L<https://sqlformat.org>

=head1 SYNOPSIS

    use strict;
    use warnings;
    use feature qw( say );

    use WebService::SQLFormat;
    my $formatter = WebService::SQLFormat->new(
        identifier_case => 'upper',
        reindent        => 1,
    );

    my $sql = shift @ARGV;

    say $formatter->format_sql($sql);

=head2 CONSTRUCTOR OPTIONS

=over 4

=item debug_level

An integer between 0 and 8.  Used to set debugging level for
L<LWP::ConsoleLogger::Easy>.  Defaults to 0.

=item identifier_case

Case to use for SQL identifiers.  One of 'upper', 'lower' or 'capitalize'.  If
no value is supplied, identifiers will not be changed.

=item keyword_case

Case to use for SQL keywords.  One of 'upper', 'lower' or 'capitalize'.  If no
value is supplied, case will not be changed.

=item reindent( 0|1)

Re-indent supplied SQL.  Defaults to 0.

=item strip_comments( 0|1 )

Remove SQL comments.  Defaults to 0.

=item ua

You may supply your own user agent.  Must be of the L<LWP::UserAgent> family.

=item url

The API url to query.  Defaults to L<https://sqlformat.org/api/v1/format>

=back

=head2 format_sql( $raw_sql )

This method expects a scalar containing the SQL which you'd like to format.
Returns the formatted SQL.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Format SQL via the sqlformat.org API",
  "AUTHOR" => "Olaf Alders <olaf\@wundercounter.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "WebService-SQLFormat",
  "EXE_FILES" => [
    "bin/sql-webformat"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "WebService::SQLFormat",
  "PREREQ_PM" => {
    "JSON::MaybeXS" => 0,
    "LWP::Protocol::https" => 0,
    "LWP::UserAgent" => 0,
    "Module::Runtime" => 0,
    "Moo" => "2.002004",
    "Types::Standard" => 0,
    "Types::URI" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test2::Bundle::More" => 0,
    "Test::RequiresInternet" => 0
  },
  "VERSION" => "0.000006",
  "test" => {
    "TESTS" => "t/*.t t/WebService/*.t"
  }
);


my %FallbackPrereqs = (
  "JSON::MaybeXS" => 0,
  "LWP::Protocol::https" => 0,
  "LWP::UserAgent" => 0,
  "Module::Runtime" => 0,
  "Moo" => "2.002004",
  "Test2::Bundle::More" => 0,
  "Test::RequiresInternet" => 0,
  "Types::Standard" => 0,
  "Types::URI" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
